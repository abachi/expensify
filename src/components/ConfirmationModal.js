import React from "react";
import Modal from "react-modal";

export const ConfirmationModal = (props) => {
  return (
    <Modal
      ariaHideApp={false}
      className="modal"
      isOpen={props.isOpen}
      onRequestClose={props.closeModal}
    >
      <div>
        <h3 className="modal__title">Do you want remove this expense?</h3>
        <div className="modal__body">
          <button
            name="cancel-btn"
            className="button button--secondary"
            onClick={props.onRequestClose}
          >
            Cancel
          </button>
          <button className="button button--danger" onClick={props.onRemove}>
            Sure
          </button>
        </div>
      </div>
    </Modal>
  );
};

export default ConfirmationModal;
