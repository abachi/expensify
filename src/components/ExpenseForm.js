import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import "react-dates/initialize";

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      amount: props.expense ? props.expense.amount : "",
      note: props.expense ? props.expense.note : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: "",
    };
  }
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    const isValid = /^\d+(\.\d{0,2})?$/.test(amount);
    if (!amount || isValid) {
      this.setState(() => ({ amount }));
    }
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onCalendarFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  handleSubmit = (e) => {
    e.preventDefault();
    const { description, note, amount, createdAt } = this.state;
    if (description.trim().length === 0 || !amount) {
      this.setState(() => ({
        error: "Please provide description and amount.",
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: description.trim(),
        amount: parseFloat(amount, 10) * 100,
        note: note.trim(),
        createdAt: createdAt.valueOf(),
      });
    }
  };

  render() {
    const {
      description,
      note,
      amount,
      error,
      createdAt,
      calendarFocused,
    } = this.state;
    return (
      <div>
        {error.length > 0 && <p>{error}</p>}
        <form onSubmit={this.handleSubmit}>
          <input
            type="text"
            name="description"
            placeholder="Description"
            value={description}
            onChange={this.onDescriptionChange}
            autoFocus
          />
          <input
            type="text"
            name="amount"
            placeholder="Amount"
            value={amount}
            onChange={this.onAmountChange}
          />
          <textarea
            placeholder="Add a note"
            value={note}
            name="note"
            onChange={this.onNoteChange}
          ></textarea>
          <SingleDatePicker
            date={createdAt}
            onDateChange={this.onDateChange}
            focused={calendarFocused}
            onFocusChange={this.onCalendarFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
