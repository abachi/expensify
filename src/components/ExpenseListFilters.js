import React from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";

import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate,
} from "../actions/filters";

export class ExpenseListFilters extends React.Component {
  state = {
    focusedInput: null,
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocus) => {
    this.setState(() => ({ calendarFocus }));
  };
  onTextSearchChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortByChange = (e) => {
    if (e.target.value === "amount") {
      this.props.sortByAmount();
    }

    if (e.target.value === "date") {
      this.props.sortByDate();
    }
  };
  render() {
    return (
      <div>
        <input
          type="text"
          name="search"
          value={this.props.filters.text}
          onChange={this.onTextSearchChange}
        />
        <select
          value={this.props.filters.sortBy}
          name="sortBy"
          onChange={this.onSortByChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.focusedInput}
          onFocusChange={this.onFocusChange}
          startDateId={"expensefy_0"}
          endDateId={"expensefy_9"}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate()),
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
});

const mapStateToProps = (state) => ({
  filters: state.filters,
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
