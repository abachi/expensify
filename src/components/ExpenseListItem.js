import React from "react";
import { Link } from "react-router-dom";

export default (props) => (
  <div>
    <span>Amount: {props.expense.amount}</span>
    <span>Description: {props.expense.description}</span>
    <span>created at: {props.expense.createdAt}</span>
    <Link to={`edit/${props.expense.id}`}>Edit</Link>
  </div>
);
