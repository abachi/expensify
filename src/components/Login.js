import React from "react";
import { connect } from "react-redux";
import { startLogin } from "../actions/auth";
import { googleAuthProvider, githubAuthProvider } from "../firebase/firebase";

export class Login extends React.Component {
  startLoginWithGoogle = () => {
    this.props.startLogin(googleAuthProvider);
  };
  startLoginWithGithub = () => {
    this.props.startLogin(githubAuthProvider);
  };

  render() {
    return (
      <div className="box-layout">
        <div className="box-layout__box">
          <h1 className="box-layout__title">Expensify</h1>
          <p>It's time to get your expenses under control</p>
          <button
            name="google-login"
            onClick={this.startLoginWithGoogle}
            className="button button--login"
          >
            Login With Google
          </button>
          <button
            name="github-login"
            onClick={this.startLoginWithGithub}
            className="button button--login"
          >
            Login With Github
          </button>
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startLogin: (provider) => dispatch(startLogin(provider)),
});

export default connect(undefined, mapDispatchToProps)(Login);
