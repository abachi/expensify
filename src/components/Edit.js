import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import ConfirmationModal from "./ConfirmationModal";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

export class Edit extends React.Component {
  state = {
    isRemoveModalOpen: false,
  };
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  };
  onRemove = () => {
    this.props.startRemoveExpense(this.props.expense);
    this.props.history.push("/");
  };
  openRemoveModal = () => {
    this.setState(() => ({ isRemoveModalOpen: true }));
  };
  closeRemoveModal = () => {
    this.setState(() => ({ isRemoveModalOpen: false }));
  };
  render() {
    return (
      <div className="page">
        <div className="page-header">
          <div className="content-container">
            <div className="page-header__body">
              <h1>Edit Expense</h1>
              <div>
                <button
                  name="remove-btn"
                  className="button button--secondary"
                  onClick={this.openRemoveModal}
                >
                  Remove Expense
                </button>
              </div>
            </div>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        </div>
        <ConfirmationModal
          isOpen={this.state.isRemoveModalOpen}
          onRequestClose={this.closeRemoveModal}
          onRemove={this.onRemove}
        />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (expense) => dispatch(startRemoveExpense(expense)),
});
const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((exp) => exp.id === props.match.params.id),
});
export default connect(mapStateToProps, mapDispatchToProps)(Edit);
