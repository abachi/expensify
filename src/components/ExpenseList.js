import React from "react";
import { connect } from "react-redux";
import numeral from "numeral";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";
import getExpensesTotal from "../selectors/expenses-total";

export const ExpenseList = (props) => {
  const total = getExpensesTotal(props.expenses);
  return (
    <div>
      {props.expenses.length === 0 ? (
        <p>No expenses</p>
      ) : (
        <div>
          <p>
            Viewing {props.expenses.length} expenses totaling{" "}
            {numeral(total).format("$0,0.00")}
          </p>
          {props.expenses.map((expense) => (
            <ExpenseListItem key={expense.id} expense={expense} />
          ))}
        </div>
      )}
    </div>
  );
};

const mapStateToProps = (state) => ({
  expenses: selectExpenses(state.expenses, state.filters),
});
export default connect(mapStateToProps)(ExpenseList);
