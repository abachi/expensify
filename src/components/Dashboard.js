import React from "react";
import ExpenseList from "./ExpenseList";
import ExpensesSummary from "./ExpensesSummary";
import ExpenseListFilters from "./ExpenseListFilters";
const Dashboard = () => (
  <div>
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);
export default Dashboard;
