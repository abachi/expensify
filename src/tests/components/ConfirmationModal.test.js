import React from "react";
import { shallow } from "enzyme";
import { ConfirmationModal } from "../../components/ConfirmationModal";

test("should render ConfirmationModal correctly", () => {
  const wrapper = shallow(<ConfirmationModal />);
  expect(wrapper).toMatchSnapshot();
});
