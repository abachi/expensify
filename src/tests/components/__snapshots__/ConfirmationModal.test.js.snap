// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ConfirmationModal correctly 1`] = `
<Modal
  ariaHideApp={false}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={0}
  isOpen={false}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <div>
    <h3
      className="modal__title"
    >
      Do you want remove this expense?
    </h3>
    <div
      className="modal__body"
    >
      <button
        className="button button--secondary"
        name="cancel-btn"
      >
        Cancel
      </button>
      <button
        className="button button--danger"
      >
        Sure
      </button>
    </div>
  </div>
</Modal>
`;
